### ISSUE TITLE
Attempting to load empty / invalid files throws PHP notices/warnings

### ISSUE THREAD

**Issue by dregad** [PHPOffice/PhpSpreadsheet#1718]
This is:

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?

Reader should check return value of failing functions (e.g. ZipArchive::open, simplexml_load_string) and throw some meaningful exception, instead of continuing.

### What is the current behavior?

Example with empty xlsx file:
```
PHP Notice:  Trying to get property 'Relationship' of non-object in /tmp/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Reader/Xlsx.php on line 342
PHP Warning:  Invalid argument supplied for foreach() in /tmp/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Reader/Xlsx.php on line 342
... etc ...
```

### What are the steps to reproduce?

Please provide a [Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

```php
<?php
include 'vendor/autoload.php';
$xlReader = new PhpOffice\PhpSpreadsheet\Reader\Xlsx();
$xlReader->load('test.xlsx');
```

Generate an empty or invalid Excel file, e.g. with `touch test.xlsx` or `echo 'hello world' >test.xlsx` and run the above script.

### Which versions of PhpSpreadsheet and PHP are affected?

Tested on PHP 7.4.3 with PhpSpreadsheet 1.15.0.

**Comment by jonnylink** [comment]
Also bumping into this issue

**Comment by SpaceInvader2020** [comment]
Have the same issue while reading just a regular xlsx file created with office 365 that has data in it, has anyone come up with the fix yet ? majority of the urls doesn't seem to be working anymore: Xlsx reader breaks with Relationship not a property error. Majority of urls used in Xlsx reader seem to be incorrect, for example instead  http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument should be http://purl.oclc.org/ooxml/officeDocument/relationships/officeDocument...

**Comment by dregad** [comment]
My original problem (which was caused by attempting to load a file which was actually not an Excel document) can easily be fixed by the following patch:

```patch
diff --git a/src/PhpSpreadsheet/Reader/Xlsx.php b/src/PhpSpreadsheet/Reader/Xlsx.php
index f07ac008..85af39da 100644
--- a/src/PhpSpreadsheet/Reader/Xlsx.php
+++ b/src/PhpSpreadsheet/Reader/Xlsx.php
@@ -328,8 +328,12 @@ class Xlsx extends BaseReader
         $unparsedLoadedData = [];
 
         $zip = new ZipArchive();
-        $zip->open($pFilename);
+        $result = $zip->open($pFilename);
+        if ($result !== true) {
+            // This does NOT catch the empty archive case - ZipArchive::open()
+            // returns true (with Deprecated notice on PHP >= 8.0)
+            throw new Exception('File "' . $pFilename . '" could not be opened (code ' . $result . ').');
+        }

         //    Read the theme first, because we need the colour scheme when reading the styles
         //~ http://schemas.openxmlformats.org/package/2006/relationships"
```

This does not handle the empty file case though, and I'm not sure about the right way to fix this. 

PHP documentation states that [libzip >= 1.6.0 treats empty archive as invalid](https://www.php.net/manual/en/ziparchive.open.php), but even though have libzip 1.7.3, ZipArchive::open('emptyfile') returns true (PHP >= 8.0 does throw a deprecated notice) so we can't rely on that.

Should load() method throw an exception given an empty file, or return a valid, empty spreadsheet object ?

I have no idea how files missing (some of) the `_rels` directories should be handled. Maybe someone with knowledge of the spec can provide some hints ?

**Comment by jonnylink** [comment]
I'd forgotten about this. We had to adopt a much more annoying tool to read files. Would be nice to see this fixed.

**Comment by oleibman** [comment]
We have just committed a change to master which is the first phase of a project to improve namespace handling. I doubt that it will fix the empty file or not-a-zip problems; I will certainly add those requirements to the next phase. I think it should handle, at least in part, the different namespace problem reported by @php-force  on April 26; please test if possible. It may or may not handle the _rels-in-the-wrong-place problem reported by @jonnylink on November 23. If it doesn't, or if you can't test against master, is it possible to upload a copy of that file so that I can analyze what is needed for it?

**Comment by jonnylink** [comment]
I'll look for one of the files and report back next week. To be honest, I'd forgotten why we couldn't use PHPOffice for certain files until the stalebot reminded me 😬

**Comment by dregad** [comment]
> I doubt that it will fix the empty file or not-a-zip problems

@oleibman I can confirm that attempting to read an **empty file** now throws *PhpOffice\PhpSpreadsheet\Reader\Exception: Could not find zip member zip://test.xlsx#_rels/.rels* instead of continuing with PHP warnings as it did until 1.18.0 included (tested today, with master branch at  bc9234e5a505e54ce00d3c815d826cd0f5a435c6), which fixes my main issue.

The **invalid file** case throws the same Reader exception (instead of *ValueError: Invalid or uninitialized Zip object* as it did before), which is good enough for me.

Looking forward to seeing this in the next official release. Thanks !

**Comment by oleibman** [comment]
Thank you for reporting the problem and confirming the fix.

**Comment by PowerKiKi** [comment]
Fixed in [1.19.0](https://github.com/PHPOffice/PhpSpreadsheet/releases/tag/1.19.0)
